### ISSUE TITLE
Several errors against large repo

### ISSUE THREAD

**Issue by d-Pixie** [codeclimate/codeclimate#23]
Fresh install of docker, boot2docker and codeclimate CLI on a MacBook with Yosemite. Project consists of PHP, JS, JSX, CSS, Less and the engines were identified as eslint, csslint and phpcodesniffer. Initial run resulted in:

    $ codeclimate analyze
    Starting analysis
    Running eslint: Done!
    error: (CC::Analyzer::Engine::EngineFailure) engine eslint failed with status 137 and stderr ""

Since the PHP code was most interesting I removed the es and css engines from the config and run again with the result of:

    $ codeclimate analyze
    Starting analysis
    Running phpcodesniffer: Done!
    error: (CC::Analyzer::Engine::EngineFailure) engine phpcodesniffer failed with status 255 and stderr "PHP Fatal error:  Uncaught exception 'Symfony\\Component\\Process\\Exception\\ProcessTimedOutException' with message 'The process \"./vendor/bin/phpcs --report=json --ignore=[test/**/*] --standard=PSR1,PSR2 /code\" exceeded the timeout of 60 sec in /usr/src/app/vendor/symfony/process/Process.php on line 1226\n"

_The list of ignored files have been replaced with the glob from the config._

The repo is rather large; 17475 PHP files, 1890 JavaScript files and 424 CSS files. It's a total of some 4.5 million lines of code (according to git stat)...

**Comment by d-Pixie** [comment]
Tried running the csslint engine as well and it errors out with:

    Running csslint: -[FATAL] failed to allocate memory

I think this is after the linting is done and it's attempting to parse the JSON. I ran it several times and got this when force quitting after it hung:

    $ codeclimate analyze
    Starting analysis
    Running csslint: Done!
    /usr/lib/ruby/gems/2.2.0/gems/json-1.8.3/lib/json/common.rb:155:in `parse': failed to allocate memory (NoMemoryError)
	from /usr/lib/ruby/gems/2.2.0/gems/json-1.8.3/lib/json/common.rb:155:in `parse'
	from /usr/src/app/lib/cc/analyzer/formatters/plain_text_formatter.rb:16:in `write'
	from /usr/src/app/lib/cc/analyzer/engine.rb:25:in `block (2 levels) in run'
	from /usr/src/app/lib/cc/analyzer/engine.rb:24:in `each_line'
	from /usr/src/app/lib/cc/analyzer/engine.rb:24:in `block in run'

My computer has about 2GB of free memory while running this so I'm assuming it's the VM that's the problem here. Any configuration options for increasing the VM memory size and see if these goes away?

**Comment by pbrisbin** [comment]
Hi @d-Pixie, sorry for the trouble here. I don't have a fix or workaround just yet, but I did want to offer some context:

- We parse each issue's JSON as it's emitted, so the fact that the error occurs in JSON parsing code doesn't mean a complete linting step was done and a separate parsing step was failing
- We intentionally run each engine in a restricted environment, which includes an artificial memory limit, so increasing available RAM on the VM or your local machine wouldn't help

One solution could be to allow users to lift the timeout and memory restrictions themselves via a command line flag. We've seen a few failures like this (for users and on the server) due to large repos hitting our limits, so we're definitely looking into solutions.

Does your repo happen to be open source? It would be great for us to test with it if possible.

**Comment by d-Pixie** [comment]
No, it's an internal project. Not on Github but using Git (in Stash at the moment). I might get permission to put it in a private repo and link that into CodeClimate if that would help?

**Comment by pbrisbin** [comment]
We'll probably be able to find or create test projects for this, if for some reason we need to look at this specific one I can let you know.

**Comment by d-Pixie** [comment]
Ok, just let me know if I can help :)

**Comment by pbrisbin** [comment]
Actually, looking again, it is the codeclimate image running out of memory, not the engine itself, so increasing available memory to your boot2docker VM could help. I believe you can tune this in the Virtual Box or VMWare GUI (whichever backend you're using).

**Comment by d-Pixie** [comment]
Ok, I'll try that tomorrow and report back :)

**Comment by d-Pixie** [comment]
A little later than tomorrow but here is an update: I changed the settings on the Virtualbox VM to allow for 4GB of memory, instead of 2Gb that is the default, and rerun the tests.

csslint passes and produced output as expected. phpcodesniffer finished but did not produce any results (and trust me, there are lots of results to produce here :)). eslint is still running, several hours later ...

So it's a mixed bag, but the initial problems seems to have been due to the VM running out of memory.
